Allow filter libraries to configure keys to ignore

prodion23 opened the pull request:
## Description
Libraries that use the filter API may want to add additional keys to the config file.  Currently the config will error when attempting to validate unexpected keys, this update will allow filter libraries to add keys that hypertrace can drop from it's own config.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

jcchavezs commented on src/hypertrace/agent/config/test_agent_config_ignored_keys.yaml:
```diff
@@ -0,0 +1,28 @@
+propagation_formats: ["B3"]
+_use_console_span_exporter: true
+enabled: false
+custom_key_for_filter_library:
```
I missed this. What is exactly this?

jcchavezs reviewed (changes requested):
On this one, I'd prefer **convention over configuration.** Allowing the user to arbitrary pass data into the config file is not only misleading for users trying to understand the configuration but also can lead to issues (imagine a user ignoring `service_name` or `reporting`). Instead I would group and document them under something specific (whose name I can't suggest because the description does not describe what is that information libraries want to pass).

prodion23 reviewed (commented):
> Instead I would group and document them under something specific (which I can't suggest because the description does not describe what is that information libraries want to pass).

In this case the goal is to allow any python library which uses the filter API to add keys to the config without hypertrace erroring during config validation.

Currently if I have a filter like: 
```python
class MyFilter():

   def __init__():
      config = read_config_file() # read the same hypertrace config.yaml
      self.some_filter_settings = config['some_custom_key']
```

hypertrace will error before my filter is initialized because it registers an errors during hypertrace config initialization: 
```
2021-08-17 22:03:50 ERROR    Unknown attribute encountered: exception=type object 'AgentConfig' has no attribute 'some_custom_key', stacktrace=Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/hypertrace/agent/config/__init__.py", line 177, in validate_config_elements
    getattr(agent_config_base, key))  # pylint: disable=W0123
AttributeError: type object 'AgentConfig' has no attribute 'some_custom_key'
```

Instead of ignoring keys we could do the following: 
```
ADDITIONAL_PERMITTED_KEYS = []

def validate_config_elements(self, config_element):
   for key in config_element:
     if key in ADDITIONAL_PERMITTED_KEYS:
       continue
```

The downside with the above is that filter library defined keys would remain accessible on the hypertrace config(which is supposed to be based on the AgentConfig proto model)

prodion23 commented:
Updated this to remove IGNORED_KEYS. 

The agent already exposes a way to skip configuration validation, but it doesn't permit `dict` type. 
I've added that here and reverted other changes.
